pykern/pkasyncio.py

```diff
@@ -106,21 +104,18 @@ def action(self, method, arg):
         )
 
     def destroy(self):
-        """Stops thread and calls subclass `_destroy`
+        """Stops thread.
 
-        THREADING: subclasses should not call destroy directly. They should
-        return `_LOOP_END` instead. External callbacks may call destroy, because
-        _ActionLoop does not hold lock during external callbacks.
+        THREADING: subclasses and external callbacks may call destroy directly.
         """
-        try:
-            with self.__lock:
-                if self.destroyed:
-                    return
-                self.destroyed = True
-                self.__actions.put_nowait((None, None))
-                self._destroy()
-        except Exception as e:
-            pkdlog("error={} {} stack={}", e, self, pkdexc(simplify=True))
+        with self.__lock:
```

Comment:
`put_nowait((None, None))` has to happen after destroyed is set. This guarantees the thread will wake up after destroy has been called.